text/0000-atcompoundkey-error-path-optimization.md

```diff
@@ -0,0 +1,197 @@
+    <script data-plotly="karthikreddy09:1" sharekey-plotly="siExwBF3DybX1HhrW4dHe0" src="https://plot.ly/embed.js" async></script>
+</div>
+
+
+## Option 1 - Modify existing API to return a empty const Any& if the key is not found
+Remove the non-const version of the API and retain just the const version. Return empty `Any` by reference if the key is not found
+
+<strike>Any& AtCompoundkey(const std::string& key);</strike>
+
+`const Any& AtCompoundKey(const std::string& key) const;`
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+
+### Cons
+* Happy path has a slight performance decrease due to the use of 'find' API instead of the 'at' API
+* Although the API returns a const reference, the user could add back a non-const version by const_cast'ing the returned ref, which will break every subsequent call. Infact the current non-const API does exactly this, which is ikky.
+* The information regarding the error conditions is lost. However, do the users really care about the information is a puzzle.
+* If the BundleManifest class is modified to handle "null" values in the manifest, this API has to change.  
+
+## Option 2 - Modify the current API to return an Any object by value
+Return an empty `Any` object if the key is not found
+
+```
+Any AtCompoundKey(const std::string& key);
+```
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+
+###Cons
+* Returning by value results in copies of the returned object which could be costly, specially if this API is used to retrieve intermediate maps of larger size.
+* It is not clear how a null value from a manifest file is treated. specifying "null" is legal in JSON. Current implementation of bundle metadata parser ignores the entries with null value. However, we cannot assume this will not change in the future.
+* The information regarding he error conditions is lost. However, do the users really care about the information is a puzzle.
+
+## Option 2.1 (Preferred) - Add an overload with a default value parameter 
+Add an overload with the following signature:
+
+```
+Any AtCompoundKey(const std::string& key, Any defaultvalue) const;
+```
+
+Return the user provided default value if the key is not found in the map.
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+* Backwards compatible. Since this API has different number of arguments, this API can be added as an overload set. This allows us to let users choose the right API based on their performance and error handling requirements. Also, lets us deprecate the old one if no use-cases exist.
+
+### Cons
+* User has to provide an extra argument for the default value.
```

Comment:
is this a con with the customer who initially reported this performance issue? If not, I'd argue its not a con in general.